src/autoschedulers/anderson2021/DefaultCostModel.cpp

```diff
@@ -0,0 +1,401 @@
+// the second moment.
+float DefaultCostModel::backprop(const Runtime::Buffer<const float> &true_runtimes, float learning_rate) {
+    internal_assert(cursor != 0);
+    internal_assert(pipeline_feat_queue.data());
+    internal_assert(schedule_feat_queue.data());
+
+    auto loss = Runtime::Buffer<float>::make_scalar();
+
+    if (!head1_filter_update.data()) {
+        auto weight_update_buffer = [](const Runtime::Buffer<float> &w) {
+            std::vector<int> size;
+            for (int i = 0; i < w.dimensions(); i++) {
+                size.push_back(w.dim(i).extent());
+            }
+            size.push_back(4);
+            auto buf = Runtime::Buffer<float>(size);
+            buf.fill(0.0f);
+            return buf;
+        };
+
+        head1_filter_update = weight_update_buffer(weights.head1_filter);
+        head1_bias_update = weight_update_buffer(weights.head1_bias);
+        head2_filter_update = weight_update_buffer(weights.head2_filter);
+        head2_bias_update = weight_update_buffer(weights.head2_bias);
+        conv1_filter_update = weight_update_buffer(weights.conv1_filter);
+        conv1_bias_update = weight_update_buffer(weights.conv1_bias);
+        timestep = 0;
+    }
+
+    Runtime::Buffer<float> dst = costs.cropped(0, 0, cursor);
+    Runtime::Buffer<float> dst_costs_per_stage = costs_per_stage.cropped({{0, cursor}, {0, num_stages}});
+
+    int fastest_idx = 0;
+    for (int i = 0; i < cursor; i++) {
+        if (true_runtimes(i) < true_runtimes(fastest_idx)) {
+            fastest_idx = i;
+        }
+    }
+
+    int result = train_cost_model(num_stages,
+                                  cursor,
+                                  num_cores,
+                                  batch_id,
+                                  pipeline_feat_queue,
+                                  schedule_feat_queue,
+                                  weights.head1_filter, weights.head1_bias,
+                                  weights.head2_filter, weights.head2_bias,
+                                  weights.conv1_filter, weights.conv1_bias,
+                                  learning_rate, timestep++,
+                                  fastest_idx,
+                                  true_runtimes.alias(),
+                                  head1_filter_update, head1_bias_update,
+                                  head2_filter_update, head2_bias_update,
+                                  conv1_filter_update, conv1_bias_update,
+                                  dst,
+                                  dst_costs_per_stage,
+                                  loss);
+    (void)result;
+    internal_assert(result == 0);
+
+    bool any_nans = false;
+    for (int i = 0; i < cursor; i++) {
+        internal_assert(cost_ptrs(i));
+        *(cost_ptrs(i)) = dst(i);
+        if (std::isnan(dst(i))) {
+            any_nans = true;
+            aslog(0) << "Prediction " << i << " is NaN. True runtime is " << true_runtimes(i) << "\n";
```

Comment:
Generally, we prefer the autoschedulers to be 'quiet' by default -- do we really want/need these to be `aslog(0)` (vs `aslog(1)` etc)?